Add first draft of new osproc.readLines

c-blake opened the pull request:
The example in the doc comment also serves as motivation (and could be converted into a test with just an `import osproc` and a `discard` of output up top).  { My actual use-case is different, but less self-contained.  With so many idle cores these days there are many circumstances where launching programs in parallel makes sense. }

The `seq[string]` return makes this new `readLines(Process)` call side-step CRLF vs LF conversion questions since `lines`-like calls in Nim all remove the ending.  So there should be no caller expectation of non-conversion (unlike immediately subsequent `execProcess`).  An `echo` loop can put back the OS-local line-endings as desired.

c-blake committed:
Add first draft of new osproc.readLines

c-blake committed:
Add test for new osproc.readLines

c-blake committed:
Rename test to start w/t to run; Also add newline to output

c-blake committed:
Suppress hint messages.

c-blake committed:
Output should match this time.

c-blake committed:
Shoulda picked a program with simpler syntax than ..lol

c-blake commented:
This seems good to go from my end.  Another application of this is with `execProcesses`, as in (fill in YADDA & BLAH as desired...):
```Nim
import osproc
const opts = {poUsePath, poStdErrToStdOut}
var nms, cmds: seq[string]
for e in YADDA:
  nms.add e
  cmds.add "BLAH " & e
let x = execProcesses(cmds, opts, n=24, afterRunEvent =
  proc(i: int, p: Process) =
    let (lines, exCode) = p.readLines ## <-- ANOTHER E.G. ##
    if lines.len > 0 or exCode != 0:
      echo nms[i], ":(exCode=", exCode, ")"
      for ln in lines: echo "  ", ln
    p.close)
quit(x)
```
While this can be implemented outside `osproc`, it applies to any situation where you want to read background process output all at once which is a very common case.  It makes sense to me for `osproc` to just provide it.

Doing it like this is also an easy way to keep output from multiple subprograms organized (rather than being all mixed together through a shared file handle).  As in this example, you can even indent the output relative to a label and conditionally output nothing (not even a "label:").

Varriount commented:
While this works for the case of a process that has a limited amount of output, it will fail for any process that runs forever. I believe it would be better if this were an iterator.

c-blake commented:
While I agree it would be bad on unbounded output, the same could be said of both `execProcess` & `execCmdEx` (as well as every backtick or `$()` command substitution ever done on Unix).  Many programs produce a deterministically small amount of output (such as one error message or not, etc.).  So, we could *also* provide a `lines` iterator, but both kinds of control flow make sense.  Do you think the iterator should return the exit code via a passed `var int` or should callers have to invoke `peekExitCode` or `waitForExit` themselves?

c-blake committed:
Address https://github.com/nim-lang/Nim/pull/15429#issuecomment-701890898
and https://github.com/nim-lang/Nim/pull/15429#issuecomment-701985976 by
factoring `readLines` into `iterator lines` and a wrapper `proc`.

c-blake commented:
Well, I've addressed @Varriount's critique and all tests still pass.  Let me know what else may matter.

Araq commented:
Needs a ``.since`` annotation.

c-blake committed:
Address https://github.com/nim-lang/Nim/pull/15429#issuecomment-702127289
and also add a `ReadIOEffect` tag to the iterator (called by the wrapper..
so it should need no separate tag, if I understand correctly).

c-blake commented:
Ok.  I also added a `ReadIOEffect` tag on the iterator.
